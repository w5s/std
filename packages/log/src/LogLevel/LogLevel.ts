import { $Object as TObject } from '@w5s/core/dist/Type/Object.js';
import { string } from '@w5s/core/dist/Type/string.js';
import { Int as TInt } from '@w5s/core/dist/Type/Int.js';
import { constant } from '@w5s/core/dist/Type/constant.js';
import type { Type } from '@w5s/core';

export const LogLevel = TObject({
  _: constant('LogLevel'),
  /**
   * Log level name
   */
  name: string,
  /**
   * Log level value (number). Often used to filter / sort logs
   */
  value: TInt,
});

export interface LogLevel extends Type.TypeOf<typeof LogLevel> {}
