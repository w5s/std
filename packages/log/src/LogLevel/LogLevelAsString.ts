import type { AsString } from '@w5s/core';
import type { LogLevel } from './LogLevel.js';

export const LogLevelAsString: AsString<LogLevel> = {
  /**
   * Returns the string representation of a level
   *
   * @example
   * ```typescript
   * const level = LogLevel('Foo', 1);
   * LogLevel.asString(level);// 'Foo'
   * ```
   * @category Formatting
   * @param self - the log level
   */
  asString(self: LogLevel): string {
    return self.name;
  },
};
