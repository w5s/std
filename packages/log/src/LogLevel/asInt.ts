import type { Int } from '@w5s/core';
import type { LogLevel } from './LogLevel.js';

/**
 * Returns the numerical representation of a level
 *
 * @example
 * ```typescript
 * const level = LogLevel('Foo', 1);
 * LogLevel.asInt(level);// 1
 * ```
 * @param level - the log level
 */
export function asInt(level: LogLevel): Int {
  return level.value;
}
