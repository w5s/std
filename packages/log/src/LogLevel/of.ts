import type { Int } from '@w5s/core';
import type { LogLevel } from '../LogLevel.js';

/**
 * Construct a new `LogLevel`
 *
 * @example
 * ```typescript
 * const level = LogLevel('UberCritical', 60);// { levelName: 'UberCritical', level: 60 }
 * ```
 * @category Constructor
 * @param name - the level string representation
 * @param value - the level value
 */
export function of(name: string, levelValue: Int): LogLevel {
  return { _: 'LogLevel', name, value: levelValue };
}
