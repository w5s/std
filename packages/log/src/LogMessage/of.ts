import type { LogMessage, LogMessageItem } from '../LogMessage.js';
import { create } from './create.js';

/**
 * Construct LogMessage from arguments
 *
 * @example
 * ```ts
 * const message = LogMessage.of('foo', ['key', 'value'], 'bar');
 * ```
 * @category Constructor
 * @param parts - constructor parameters
 */
export function of(...parts: Array<LogMessageItem>): LogMessage {
  return create(parts);
}
