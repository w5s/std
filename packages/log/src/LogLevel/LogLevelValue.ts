import { IntBounded } from '@w5s/core/dist/Int/IntBounded.js';
import type { Int } from '@w5s/core';
import { of } from './of.js';

const None = of('none', IntBounded.minValue);
/**
 * Critical log level (50)
 */
const Critical = of('critical', 50 as Int);
/**
 * Error log level (40)
 */
const Error = of('error', 40 as Int);
/**
 * Warning log level (30)
 */
const Warning = of('warning', 30 as Int);
/**
 * Info log level (20)
 */
const Info = of('info', 20 as Int);
/**
 * Debug log level (10)
 */
const Debug = of('debug', 10 as Int);

/**
 * @namespace
 */
export const LogLevelValue = {
  None,
  Critical,
  Error,
  Warning,
  Info,
  Debug,
};

export type LogLevelValue = Lowercase<keyof typeof LogLevelValue>;
