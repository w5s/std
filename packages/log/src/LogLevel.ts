import { Callable } from '@w5s/core/dist/Callable.js';
import { LogLevel as LogLevelType } from './LogLevel/LogLevel.js';
import { LogLevelComparable } from './LogLevel/LogComparable.js';
import { of } from './LogLevel/of.js';
import { LogLevelValue } from './LogLevel/LogLevelValue.js';
import { match } from './LogLevel/match.js';
import { from } from './LogLevel/from.js';

export interface LogLevel extends LogLevelType {}

/**
 * @example
 * @namespace
 */
export const LogLevel = Callable({
  [Callable.symbol]: of,
  ...LogLevelType,
  ...LogLevelComparable,
  ...LogLevelValue,
  of,
  from,
  /**
   * Returns the string representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.format(level);// 'Foo'
   * ```
   * @param level - the log level
   */
  format(level: LogLevel): string {
    return level.name;
  },

  /**
   * Returns the numerical representation of a level
   *
   * @example
   * ```ts
   * const level = LogLevel('Foo', 1);
   * LogLevel.value(level);// 1
   * ```
   * @param level - the log level
   */
  value(level: LogLevel): number {
    return level.value;
  },
  match,
});
