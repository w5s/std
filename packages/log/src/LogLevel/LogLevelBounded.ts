import type { Bounded } from '@w5s/core';
import { LogLevelValue } from './LogLevelValue.js';
import type { LogLevel } from './LogLevel.js';

export const LogLevelBounded: Bounded<LogLevel> = {
  minValue: LogLevelValue.None,
  maxValue: LogLevelValue.Critical,
};
