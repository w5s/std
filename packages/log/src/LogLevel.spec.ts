import { describe, it, expect } from 'vitest';
import { LogLevel } from './LogLevel.js';
import { LogLevelBounded } from './LogLevel/LogLevelBounded.js';
import { LogLevelComparable } from './LogLevel/LogLevelComparable.js';
import { LogLevelValue } from './LogLevel/LogLevelValue.js';
import { of } from './LogLevel/of.js';
import { LogLevelAsString } from './LogLevel/LogLevelAsString.js';
import { asInt } from './LogLevel/asInt.js';

describe('LogLevel', () => {
  it('is an alias to functions', () => {
    expect(LogLevel).toEqual(expect.objectContaining(LogLevelBounded));
    expect(LogLevel).toEqual(expect.objectContaining(LogLevelComparable));
    expect(LogLevel).toEqual(expect.objectContaining(LogLevelValue));
    expect(LogLevel).toEqual(expect.objectContaining(LogLevelAsString));
    expect(LogLevel).toEqual(
      expect.objectContaining({
        of,
        asInt,
      }),
    );
  });
});
