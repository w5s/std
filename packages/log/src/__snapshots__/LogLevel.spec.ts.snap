// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LogLevel > Critical > should have correct value > Critical 1`] = `
{
  "_": "LogLevel",
  "name": "critical",
  "value": 50,
}
`;

exports[`LogLevel > Debug > should have correct value > Debug 1`] = `
{
  "_": "LogLevel",
  "name": "debug",
  "value": 10,
}
`;

exports[`LogLevel > Error > should have correct value > Error 1`] = `
{
  "_": "LogLevel",
  "name": "error",
  "value": 40,
}
`;

exports[`LogLevel > Info > should have correct value > Info 1`] = `
{
  "_": "LogLevel",
  "name": "info",
  "value": 20,
}
`;

exports[`LogLevel > Warning > should have correct value > Warning 1`] = `
{
  "_": "LogLevel",
  "name": "warning",
  "value": 30,
}
`;
